File: api/tts.js
// api/tts.js — Hugging Face TTS (free, no credit card)
// Arabic : facebook/mms-tts-ara
// English: facebook/mms-tts-eng

const MODEL_AR = 'facebook/mms-tts-ara';
const MODEL_EN = 'facebook/mms-tts-eng';
const HF_BASE  = 'https://api-inference.huggingface.co/models/';

export default async function handler(req, res) {
  if (req.method === 'OPTIONS') {
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
    return res.status(204).end();
  }

  if (req.method !== 'POST') return res.status(405).json({ error: 'Method not allowed' });

  const { text, lang } = req.body;
  if (!text?.trim()) return res.status(400).json({ error: 'Missing text' });

  const apiKey = process.env.HUGGINGFACE_API_KEY;
  if (!apiKey) return res.status(500).json({ error: 'HUGGINGFACE_API_KEY not set' });

  const model = lang === 'ar' ? MODEL_AR : MODEL_EN;

  try {
    const hfRes = await fetch(HF_BASE + model, {
      method:  'POST',
      headers: {
        'Authorization': `Bearer ${apiKey}`,
        'Content-Type':  'application/json'
      },
      body: JSON.stringify({ inputs: text.trim().slice(0, 2500) })
    });

    if (!hfRes.ok) {
      const err = await hfRes.text();
      return res.status(502).json({ error: `HF error ${hfRes.status}`, detail: err.slice(0, 200) });
    }

    const audio       = await hfRes.arrayBuffer();
    const contentType = hfRes.headers.get('content-type') || 'audio/flac';

    res.setHeader('Content-Type', contentType);
    res.setHeader('Cache-Control', 'no-cache');
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.status(200).send(Buffer.from(audio));

  } catch (err) {
    res.status(500).json({ error: err.message });
  }
}

File: api/tts.py
# api/tts.py — Microsoft Edge TTS (edge-tts)
# Completely free, no API key, no account needed
# Arabic: ar-SA-HamedNeural (male) / ar-SA-ZariyahNeural (female)
# English: en-US-ChristopherNeural (male) / en-US-JennyNeural (female)

from http.server import BaseHTTPRequestHandler
import json
import asyncio
import edge_tts

VOICE_EN = 'en-US-ChristopherNeural'
VOICE_AR = 'ar-SA-HamedNeural'

async def synthesize(text, voice):
    communicate = edge_tts.Communicate(text, voice)
    audio = b''
    async for chunk in communicate.stream():
        if chunk['type'] == 'audio':
            audio += chunk['data']
    return audio

class handler(BaseHTTPRequestHandler):

    def log_message(self, format, *args):
        pass  # suppress access logs

    def do_OPTIONS(self):
        self.send_response(204)
        self._cors()
        self.end_headers()

    def do_POST(self):
        try:
            length = int(self.headers.get('Content-Length', 0))
            body   = json.loads(self.rfile.read(length))
            text   = body.get('text', '').strip()[:2500]
            lang   = body.get('lang', 'en')

            if not text:
                self.send_response(400)
                self._cors()
                self.send_header('Content-Type', 'application/json')
                self.end_headers()
                self.wfile.write(b'{"error":"Missing text"}')
                return

            voice = VOICE_AR if lang == 'ar' else VOICE_EN
            audio = asyncio.run(synthesize(text, voice))

            self.send_response(200)
            self._cors()
            self.send_header('Content-Type', 'audio/mpeg')
            self.send_header('Cache-Control', 'no-cache')
            self.end_headers()
            self.wfile.write(audio)

        except Exception as e:
            self.send_response(500)
            self._cors()
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({'error': str(e)}).encode())

    def _cors(self):
        self.send_header('Access-Control-Allow-Origin',  '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
